samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/util/ExtendedTimeValidator.java

```diff
@@ -0,0 +1,408 @@
+                    isEntryRetractEarlierThanAvailable = true;
+                    valid = false;
+                }
+
+                // Retract date can't be before due date; push it to the due date
+                if( !isEntryRetractEarlierThanAvailable && (retractDate != null && dueDate != null && retractDate.before( dueDate )) )
+                {
+                    entry.setRetractDate( dueDate );
+                }
+            }
+
+            // Due date can't be the same as start date
+            if( dueDate != null && startDate != null && dueDate.equals( startDate ) )
+            {
+                String errorMsg = getError( ERROR_KEY_DUE_SAME_AS_START, entry );
+                context.addMessage( null, new FacesMessage( FacesMessage.SEVERITY_WARN, errorMsg, null ) );
+                valid = false;
+            }
+
+            // If time limit is set, ensure open window is not less than the time limit
+            boolean hasTimer = (entry.getTimeHours() != null && entry.getTimeHours() > 0) || (entry.getTimeMinutes() != null && entry.getTimeMinutes() > 0);
+            if (hasTimer)
+            {
+                Date due = entry.getRetractDate() != null ? entry.getRetractDate() : entry.getDueDate();
+                long timerMinutes = 0;
+                if( entry.getTimeHours() != null )
+                {
+                    timerMinutes += entry.getTimeHours().longValue() * 60;
+                }
+                if( entry.getTimeMinutes() != null )
+                {
+                    timerMinutes += entry.getTimeMinutes().longValue();
+                }
+
+                long openWindowMinutes = (due.getTime() - entry.getStartDate().getTime()) / 1000 / 60;
+                if( openWindowMinutes < timerMinutes )
+                {
+                    String errorString = getError( ERROR_KEY_OPEN_WINDOW_LESS_THAN_LIMIT, entry );
+                    context.addMessage( null, new FacesMessage( FacesMessage.SEVERITY_WARN, errorString, null ) );
+                    valid = false;
+                }
+            }
+        }
+
+        // Check for duplicate users
+        Set<String> duplicateUsers = findDuplicates(users);
+        if( !duplicateUsers.isEmpty() )
+        {
+            String dupUsers = "";
+            int count = 0;
+            int end = users.size();
+            for( String entry : duplicateUsers )
+            {
+                if( count == 0 )
+                {
+                    dupUsers = "'" + getUserName( entry ) + "'";
+                }
+                else if( count < (end - 1) )
+                {
+                    dupUsers = dupUsers + ", '" + getUserName( entry ) + "'";
+                }
+                else
+                {
+                    String and = ContextUtil.getLocalizedString( ASSESSMENT_SETTINGS_BUNDLE, MSG_KEY_AND );
+                    dupUsers = dupUsers + ", " + and + " '" + getUserName( entry );
```

Comment:
wow this is a lot of lifting for duplicates!

Reply:
Yeah I didn't write most of this code. Some could certainly be improved but I tried not to touch original logic.